### ISSUE TITLE
Add hasFirstMember property

### ISSUE THREAD

**Issue by marksem** [semanticarts/gist#261]
It would seem useful to have a gist:hasFirstMember property, that is a rdfs:subPropertyOf gist:hasOrderedMember.

Then it is easy to find the first member of an ordered list, after which you can follow gist:directlyPrecedes to walk the list.

**Comment by uscholdm** [comment]
Makes sense in principle. Is there a use case that we know of?  I don't think this ever got much use anyway.   The very idea of a class for OrderedMember is a but clunky (I'm dissing my own creation, BTW).

**Comment by marksem** [comment]
We added hasFirstMember  at MS for modeling Kill Chains as ordered lists of techniques.
________________________________

**Comment by uscholdm** [comment]
Hah – so we did.  Generally, one use at one client is not justification for inclusion in gist – that keeps it small.  I kind of like it, but have been trained by day to say when in doubt, leave it out.

M.

**Comment by daliadahleh** [comment]
I used OrderedCollection for chaining stored procedures in the DCA. To retrieve the collection in order, I queried for the member that doesn't have another member that gist:directlyPrecedes it, then moved down the chain from there. But it may be simpler and/or faster to explicitly point to the first member.

Also, to validate that an OrderedCollection is a proper chain, I have a SHACL shape that uses SHACL-SPARQL to verify that there is only one head and one tail. If it had its own property, that could simplify the shape definition so that it doesn't require a query.

**Comment by uscholdm** [comment]
I love it. Sounds like a good argument for adding hasFirstMember.  It is functional, of course (in both senses of the word).

**Comment by rjyounes** [comment]
Dalia wrote: "To retrieve the collection in order, I queried for the member that doesn't have another member that gist:directlyPrecedes it."

This was going to be my response. hasFirstMember is redundant. Then why not hasLastMember? hasSecondMember (OK, that one's going a bit too far but I add it as a reductio ad absurdum).

**Comment by daliadahleh** [comment]
> hasFirstMember is redundant. Then why not hasLastMember? hasSecondMember (OK, that one's going a bit too far but I add it as a reductio ad absurdum).

There is arguably some redundancy already between hasOrderedMember and directlyPrecedes. For no redundancy at all, we could have _only_ hasFirstMember and directlyPrecedes (i.e. no need for the collection to point to every member when they already point to each other).

**Comment by marksem** [comment]
> hasFirstMember is redundant. Then why not hasLastMember? hasSecondMember (OK, that one's going a bit too far but I add it as a reductio ad absurdum).

Complete lack of redundancy needn't be the highest goal.  I believe usability should be.  And to me this means the common use cases.  The modern philosophy of "model your queries" over "third normal form" might apply here.  

So yes, hasFirstMember is redundant with finding the one that has to predecessor.  But the use case of finding the first one seems so common, it seems like it is useful to have an easy pointer to that.
Traditionally, linked lists have a "first member" pointer, because it is such common use case to get to the first thing and then go to the "next" thing, etc.

**Comment by marksem** [comment]
> There is arguably some redundancy already between hasOrderedMember and directlyPrecedes. For no redundancy at all, we could have only hasFirstMember and directlyPrecedes (i.e. no need for the collection to point to every member when they already point to each other).

Thoughts on minimalism for Collections:

A basic Collection needs hasMember or memberOf.  I can see no way around it since that is the essence of a collection.  (Barring RDF lists and sequences, which are much more complicated I think.) 
An OrderedCollection, being a collection, will inherit this too (unless we're willing to have ordered collections to be significantly different from regular Collections). So let's assume Ordered Collections will have a pointer to every member.  After that, the minimal structure for an ordered list could either be:

 1) adding a gist:sequence value to every member
or
 2) every member (but the last) has directlyPrecedes.  

Problems with #1
 - an item can't be in more than 1 collection (won't know which gist:sequence goes with which collection)
Problems with #2
 - sorting the list is impossible(?) in a SPARQL query
 - an item can't really be in more than 1 collection (won't know which directlyPrecedes goes with which collection)

So #1 seems better (less problems), but still, member items can be in only 1 collection.  Not a very realistic constraint in the real world (i.e. our clients).  

So a more usable Ordered Collection concept, allowing things to be in more than 1 collection, would seem to need a concept of "slots", as in OLO (Ordered list ontology [1]).  In that case, may be best to simply adopt OLO since it only has 2 classes and 7 properties, some of which would replace a few gist ones.

[1] http://smiy.sourceforge.net/olo/spec/orderedlistontology.html

**Comment by daliadahleh** [comment]
I agree that minimalism shouldn't be the only goal.

> sorting the list is impossible(?) in a SPARQL query

Not impossible :) Here you go:

```
select ?member (count(?distance) as ?order) where {
  ?first gist:directlyPrecedes* ?distance .
  ?distance gist:directlyPrecedes* ?member .
  filter not exists {?any gist:directlyPrecedes ?first}
}
group by ?member order by ?order
```

>an item can't really be in more than 1 collection

To have a member be in multiple collections, I used:

`:_Chain1_Member1 :isProxyFor :_Member1 and :directlyPrecedes _Chain1_Member2`

This only adds one instance and one property per member per chain. It's similar in principle to OLO, but what I don't like about OLO is that it forces you to use a numbered sequence (olo:index is mandatory, and olo:next is optional).

> A basic Collection needs hasMember or memberOf. I can see no way around it since that is the essence of a collection.

Linked lists generally only point to the first node.

**Comment by rjyounes** [comment]
@daliadahleh yes, you have to use proxies for an object to be in more than one collection. I would like to see gist provide that functionality. There is discussion of this in #112.

**Comment by marksem** [comment]
Closed by #243
